torch/_inductor/codegen/cuda/gemm_template.py

```diff
@@ -0,0 +1,496 @@
+        if op.gemm_kind == cutlass_lib.GemmKind.Universal3x:
+            epilogue_schedule_str = str(op.epilogue_schedule).split(".")[-1]
+            result = epilogue_schedule_str.lower().startswith("tma")
+        return result
+
+    @staticmethod
+    def define_gemm_instance(
+        op: cutlass_gemm_op.GemmOperation,
+    ) -> Tuple[str, str]:
+        if op.gemm_kind == cutlass_lib.GemmKind.Universal3x:
+            emitter = cutlass_gemm_op.EmitGemmUniversal3xInstance()
+            op_def = emitter.emit(op)
+            pattern = re.compile(r"\s*struct\s(.*?)\s:")
+            decl = [line for line in op_def.split("\n") if "struct " in line][-1]
+        else:
+            emitter = cutlass_gemm_op.EmitGemmInstance()
+            op_def = emitter.emit(op)
+            op_def = op_def.replace(
+                "cutlass::gemm::device::Gemm", "cutlass::gemm::device::GemmUniversal"
+            )
+            op_def = op_def.replace("false,", "")
+            pattern = re.compile(r"\s*using\s(.*?)\s=")
+            decl = op_def.split("\n")[2]
+        match = pattern.match(decl)
+        if match is None:
+            raise RuntimeError("Invalid Gemm config: \n" + op_def)
+        op_type = match.groups()[0]
+        if op.gemm_kind == cutlass_lib.GemmKind.Universal3x:
+            op_def += f"\n  using {op_type}_device_type = cutlass::gemm::device::GemmUniversalAdapter<{op_type}>;\n"
+            op_type = f"{op_type}_device_type"
+        return op_def, op_type
+
+    @staticmethod
+    def should_swap_XW(
+        bias: IRNode,
+        beta: float,
+    ) -> bool:
+        return True
+
+        # TODO(ipiszy): Check whether it's necessary to swap X/W.
+        # strides = bias.get_stride()
+        # if strides[-1] != 1:
+        #     return True
+        # for stride in strides[:-1]:
+        #     if stride != 0:
+        #         return True
+        # return False
+
+    @staticmethod
+    def swap_XW(op: cutlass_gemm_op.GemmOperation) -> cutlass_gemm_op.GemmOperation:
+        # Swap X and W in GemmOperation.
+        new_op = copy.deepcopy(op)
+        new_op.A.layout = CUTLASSGemmTemplate.flip_cutlass_layout(new_op.A.layout)
+        new_op.B.layout = CUTLASSGemmTemplate.flip_cutlass_layout(new_op.B.layout)
+        new_op.A, new_op.B = new_op.B, new_op.A
+        new_op.C.layout = CUTLASSGemmTemplate.flip_cutlass_layout(new_op.C.layout)
+        new_op.D.layout = CUTLASSGemmTemplate.flip_cutlass_layout(new_op.D.layout)
+        return new_op
+
+    def filter_op(
+        self,
+        op: cutlass_gemm_op.GemmOperation,
+    ) -> cutlass_gemm_op.GemmOperation:
+        # Skip simt kernels
+        if (
+            op.tile_description.math_instruction.opcode_class
+            == cutlass_lib.OpcodeClass.Simt
+        ):
+            return None
+
+        # Skip sparse and grouped kernels
+        if op.gemm_kind in {
+            cutlass_lib.GemmKind.Sparse,
+            cutlass_lib.GemmKind.Grouped,
+        }:
```

Comment:
Should we instead explicitly list the types we support (and return otherwise)? As written, this may break if a new `cutlass_lib.GemmKind` is added.

Reply:
The existing implementation copies from AIT and is very hacky. It actually relies on GemmKind like Planar, Complex, etc, which doesn't seem make sense. However if I only use GemmKind.Gemm there is no config left. So for now I'd like to keep it as it is. CUTLASS folks suggest using the new Python cutlass interface to generate configs, which might be a follow-up.

Reply:
I believe, AIT uses an allow-list approach accepting only `GemmKind.Universal` and `GemmKind.Universal3x`: https://github.com/facebookincubator/AITemplate/blob/main/python/aitemplate/backend/cuda/gemm_universal/common.py#L823-L826 . I guess, this should work here, too? `GemmKind.Gemm` must be something pre-`GemmUniversal`.

Reply:
I see, I tried "GemmKind.Gemm" because we are replacing Gemm with GemmUniversal in generated code. But GemmKind.Gemm doesn't work well. Now I see that AIT actually uses GemmKind.GemmUniversal to emit GemmInstance, which I missed before. Thanks for pointing this out!